Sources/Sentry/SentryDevice.mm

```diff
@@ -0,0 +1,266 @@
+ *   - 2020 iMac (x86_64h)
+ *   - 2021 MBP (M1 reported as arm64e)
+ *   - iPhone simulators on all of those macs
+ *   - iPhone 13 mini (arm64e)
+ *   - iPod Touch (6th gen) (armv8)
+ * @endcode
+ */
+NSString *
+getCPUType(NSNumber *_Nullable subtype)
+{
+    cpu_type_t type;
+    size_t typeSize = sizeof(type);
+    if (SENTRY_LOG_ERRNO(sysctlbyname("hw.cputype", &type, &typeSize, NULL, 0)) != 0) {
+        if (subtype != nil) {
+            return
+                [NSString stringWithFormat:@"no CPU type for unknown subtype %d", subtype.intValue];
+        }
+        return @"no CPU type or subtype";
+    }
+    switch (type) {
+    default:
+        if (subtype != nil) {
+            return [NSMutableString
+                stringWithFormat:@"unknown CPU type (%d) and subtype (%d)", type, subtype.intValue];
+        }
+        return [NSMutableString stringWithFormat:@"unknown CPU type (%d)", type];
+    case CPU_TYPE_X86_64:
+        // I haven't observed this branch being taken for 64-bit x86 architectures. Rather, the
+        // x86 branch is taken, and then the subtype is reported as the 64-bit
+        // subtype. Tested on a 2020 Intel-based iMac and 2015 MBP. (armcknight 21 Sep 2022)
+        return @"x86_64";
+    case CPU_TYPE_X86:
+        return @"x86";
+    case CPU_TYPE_ARM:
+        return @"arm";
+    case CPU_TYPE_ARM64:
+        return @"arm64";
+    case CPU_TYPE_ARM64_32:
+        return @"arm64_32";
+    }
+}
+} // namespace
+
+NSString *
+sentry_getCPUArchitecture(void)
+{
+    cpu_subtype_t subtype;
+    size_t subtypeSize = sizeof(subtype);
+    if (SENTRY_LOG_ERRNO(sysctlbyname("hw.cpusubtype", &subtype, &subtypeSize, NULL, 0)) != 0) {
+        return getCPUType(nil);
+    }
+    switch (subtype) {
+    default:
+        return getCPUType(@(subtype));
+    case CPU_SUBTYPE_X86_64_H:
+        return @"x86_64h";
+    case CPU_SUBTYPE_X86_64_ALL:
+        return @"x86_64";
+    case CPU_SUBTYPE_ARM_V6:
+        return @"armv6";
+    case CPU_SUBTYPE_ARM_V7:
+        return @"armv7";
+    case CPU_SUBTYPE_ARM_V7S:
+        return @"armv7s";
+    case CPU_SUBTYPE_ARM_V7K:
+        return @"armv7k";
+    case CPU_SUBTYPE_ARM64_V8:
+        // this also catches CPU_SUBTYPE_ARM64_32_V8 since they are both defined as
+        // ((cpu_subtype_t) 1)
+        return @"armv8";
+    case CPU_SUBTYPE_ARM64E:
+        return @"arm64e";
+    }
+}
+
+NSString *
+sentry_getOSName(void)
+{
+#if TARGET_OS_MACCATALYST
+    return @"Catalyst";
+#elif SENTRY_HAS_UIKIT
+    return UIDevice.currentDevice.systemName;
+#else
+    return @"macOS";
+#endif // SENTRY_HAS_UIKIT
+}
+
+NSString *
+sentry_getOSVersion(void)
+{
+#if TARGET_OS_WATCH
+    return WKInterfaceDevice.currentDevice.systemVersion;
+#elif SENTRY_HAS_UIKIT
+    return UIDevice.currentDevice.systemVersion;
+#else
+    // based off of
+    // https://github.com/lmirosevic/GBDeviceInfo/blob/98dd3c75bb0e1f87f3e0fd909e52dcf0da4aa47d/GBDeviceInfo/GBDeviceInfo_OSX.m#L107-L133
```

Comment:
Funny, lmirosevic was a lecturer for me in one course at my university many years ago 😁.

Reply:
😮  small world!